### PR TITLE
Separate Invariants for AIFS-ENS

### PR DESCRIPTION
**Pull request by mariusaurus**
# Earth2Studio Pull Request

## Description
This PR removes static variables from set of input variables and downloads them when loading the model package instead. Additionally, this enables AIFS-ENS to be initialised with CDS or NCAR data sources. So far, data sources are limited to IFS and AIFS.
The tests for the model have been adapted accordingly.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.
- [x] Assess and address Greptile feedback (AI code review bot for guidance; use discretion, addressing all feedback is not required).

### CODE REVIEW

**Comment by NickGeneva** [comment]
Comparison between the previous version and this branch seems reasonable close... (kinda)

image


using the simple inference script:

```python
from earth2studio.models.px import AIFSENS
from earth2studio.data import IFS, fetch_data

import torch
from datetime import datetime, timedelta
import numpy as np

time = np.array([datetime(2026, 1, 1, 0)], dtype=np.datetime64)

seed = 0
np.random.seed(seed)
torch.manual_seed(seed)
if torch.cuda.is_available():
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)


model = AIFSENS.load_model(AIFSENS.load_default_package()).to("cuda:0")
data = IFS()

x, coords = fetch_data(
        source=data,
        time=time,
        lead_time=model.input_coords()["lead_time"],
        variable=model.input_coords()["variable"],
        device="cuda:0",
)
x_out, coords = model(x, coords)

torch.save(x_out.cpu(), "aifs_ens_mkoch.pt")
np.save( "aifs_variable_mkoch.npy", coords['variable'])
```

<details>

<summary>Plot Script</summary>

```python
import os
import numpy as np
import torch
import matplotlib.pyplot as plt

# --------------------
# Config
# --------------------
OLD_PT = "aifs_ens_old.pt"
MK_PT  = "aifs_ens_mkoch.pt"
OLD_VARS = "aifs_variable_old.npy"
MK_VARS  = "aifs_variable_mkoch.npy"

# Choose indices to visualize
TIME_IDX = 0
LEAD_IDX = 0

# Plotting options
CMAP = "RdBu_r"
DIFF_CMAP = "PiYG"
OUTPUT = "aifs_ens_compare.jpg"

old_tensor = torch.load(OLD_PT, map_location="cpu")
mk_tensor  = torch.load(MK_PT, map_location="cpu")

old_np = old_tensor.detach().cpu().numpy()
mk_np  = mk_tensor.detach().cpu().numpy()

# Load variable vocabularies
old_vars = np.load(OLD_VARS, allow_pickle=True)
mk_vars  = np.load(MK_VARS, allow_pickle=True)

# Ensure they are lists of strings
old_vars = [str(v) for v in old_vars.tolist()]
mk_vars  = [str(v) for v in mk_vars.tolist()]
common_vars = [v for v in mk_vars if v in old_vars]

# Map variable name to index
old_index = {v: i for i, v in enumerate(old_vars)}
mk_index  = {v: i for i, v in enumerate(mk_vars)}

T_old, L_old, _, Y_old, X_old = old_np.shape
T_mk,  L_mk,  _, Y_mk,  X_mk  = mk_np.shape


n = len(common_vars)
ncols = 3
nrows = n
fig, axs = plt.subplots(nrows=nrows, ncols=ncols, figsize=(12, 3.5 * nrows))

# Ensure axs is 2D
axs = np.atleast_2d(axs)

for i, v in enumerate(common_vars):
    ov = old_np[TIME_IDX, LEAD_IDX, old_index[v], :, :]
    mv = mk_np[TIME_IDX, LEAD_IDX, mk_index[v],  :, :]

    # Robust color scale for the two fields together
    stack = np.concatenate([ov.ravel(), mv.ravel()])
    stack = stack[np.isfinite(stack)]
    if stack.size > 0:
        vmin, vmax = np.nanpercentile(stack, [2, 98])
    else:
        vmin, vmax = np.nanmin(ov), np.nanmax(ov)

    # Diff symmetric scale
    diff = mv - ov
    dstack = diff[np.isfinite(diff)]
    if dstack.size > 0:
        dmin, dmax = np.nanpercentile(dstack, [2, 98])
        dabs = max(abs(dmin), abs(dmax))
    else:
        dabs = max(abs(np.nanmin(diff)), abs(np.nanmax(diff)))

    # OLD
    ax = axs[i, 0]
    im0 = ax.imshow(ov, cmap=CMAP, vmin=vmin, vmax=vmax, origin="upper")
    ax.set_title(f"OLD {v} [t={TIME_IDX}, l={LEAD_IDX}]")
    ax.set_xticks([]); ax.set_yticks([])
    fig.colorbar(im0, ax=ax, orientation="horizontal", fraction=0.05, pad=0.06)

    # MK
    ax = axs[i, 1]
    im1 = ax.imshow(mv, cmap=CMAP, vmin=vmin, vmax=vmax, origin="upper")
    ax.set_title(f"MK {v} [t={TIME_IDX}, l={LEAD_IDX}]")
    ax.set_xticks([]); ax.set_yticks([])
    fig.colorbar(im1, ax=ax, orientation="horizontal", fraction=0.05, pad=0.06)

    # DIFF
    ax = axs[i, 2]
    im2 = ax.imshow(diff, cmap="PiYG", vmin=-dabs, vmax=dabs, origin="upper")
    ax.set_title(f"Δ (MK − OLD) {v}")
    ax.set_xticks([]); ax.set_yticks([])
    fig.colorbar(im2, ax=ax, orientation="horizontal", fraction=0.05, pad=0.06)

plt.tight_layout()
plt.savefig("aifs_ens_compare_vars.jpg", dpi=150)
print("Saved to aifs_ens_compare_vars.jpg")
```

</detials>

**Comment by NickGeneva** [comment]
Update:

Okay, so I wanted to check to see if that difference was due to some internal randomness in anemoi. So I dumped the input to the model right before the anemoi `predict_step` call.

The tensor shape here is `torch.Size([1, 2, 542080, 101])` with the last one being the variable dim and the results are:

<details>

```
tensor(0.) q50
tensor(0.) q100
tensor(0.) q150
tensor(0.) q200
tensor(0.) q250
tensor(0.) q300
tensor(0.) q400
tensor(0.) q500
tensor(0.) q600
tensor(0.) q700
tensor(0.) q850
tensor(0.) q925
tensor(0.) q1000
tensor(0.) t50
tensor(0.) t100
tensor(0.) t150
tensor(0.) t200
tensor(0.) t250
tensor(0.) t300
tensor(0.) t400
tensor(0.) t500
tensor(0.) t600
tensor(0.) t700
tensor(0.) t850
tensor(0.) t925
tensor(0.) t1000
tensor(0.) u50
tensor(0.) u100
tensor(0.) u150
tensor(0.) u200
tensor(0.) u250
tensor(0.) u300
tensor(0.) u400
tensor(0.) u500
tensor(0.) u600
tensor(0.) u700
tensor(0.) u850
tensor(0.) u925
tensor(0.) u1000
tensor(0.) v50
tensor(0.) v100
tensor(0.) v150
tensor(0.) v200
tensor(0.) v250
tensor(0.) v300
tensor(0.) v400
tensor(0.) v500
tensor(0.) v600
tensor(0.) v700
tensor(0.) v850
tensor(0.) v925
tensor(0.) v1000
tensor(0.) w50
tensor(0.) w100
tensor(0.) w150
tensor(0.) w200
tensor(0.) w250
tensor(0.) w300
tensor(0.) w400
tensor(0.) w500
tensor(0.) w600
tensor(0.) w700
tensor(0.) w850
tensor(0.) w925
tensor(0.) w1000
tensor(0.) z50
tensor(0.) z100
tensor(0.) z150
tensor(0.) z200
tensor(0.) z250
tensor(0.) z300
tensor(0.) z400
tensor(0.) z500
tensor(0.) z600
tensor(0.) z700
tensor(0.) z850
tensor(0.) z925
tensor(0.) z1000
tensor(0.) u10m
tensor(0.) v10m
tensor(0.) d2m
tensor(0.) t2m
tensor(-413.1261) lsm
tensor(0.) msl
tensor(-6139072.5000) sdor
tensor(0.) skt
tensor(-338.4066) slor
tensor(0.) sp
tensor(0.) tcw
tensor(-270553.7188) z
tensor(0.) cp06
tensor(0.) tp06
tensor(0.) cos_latitude
tensor(0.) cos_longitude
tensor(0.) sin_latitude
tensor(0.) sin_longitude
tensor(0.) cos_julian_day
tensor(0.) cos_local_time
tensor(0.) sin_julian_day
tensor(0.) sin_local_time
tensor(0.) insolation
```

</details>

So there is some difference between the NCAR and the IFS invariants... replacing the invariant tensor with:

```python
ifs = IFS(cache=True, verbose=False)
  x_inv, coords_inv = fetch_data(
      source=ifs,
      time=np.array([np.datetime64("2026-01-01T00:00:00")]),
      variable=["lsm", "sdor", "slor", "z"],
  )
```

Makes everything fully consistent. However this appears to basically have no impact on the actual prediction... both produce nearly the same error ranges.

**Comment by NickGeneva** [comment]
Alright, trimmed up this PR a bit.

For documentation purposes I used the following inference code to produce a set of output tensor files using current main branch and this branch.

```python
from earth2studio.models.px import AIFSENS
from earth2studio.data import IFS, fetch_data

import torch
from datetime import datetime, timedelta
import numpy as np

time = np.array([datetime(2026, 1, 1, 0)], dtype=np.datetime64)

seed = 0
np.random.seed(seed)
torch.manual_seed(seed)
if torch.cuda.is_available():
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)


model = AIFSENS.load_model(AIFSENS.load_default_package()).to("cuda:0")
data = IFS()

x, coords = fetch_data(
        source=data,
        time=time,
        lead_time=model.input_coords()["lead_time"],
        variable=model.input_coords()["variable"],
        device="cuda:0",
)
x_out, coords = model(x, coords)

torch.save(x_out.cpu(), "aifs_ens_test.pt")
np.save( "aifs_variable_ngeneva.npy", coords['variable'])
```

I then ran the following script to product a comparion plot

<details>

```python
import os
import numpy as np
import torch
import matplotlib.pyplot as plt

# --------------------
# Config
# --------------------
OLD_PT = "aifs_ens_old.pt"
MK_PT  = "aifs_ens_test.pt"
OLD_VARS = "aifs_variable_old.npy"
MK_VARS  = "aifs_variable_ngeneva.npy"

# Choose indices to visualize
TIME_IDX = 0
LEAD_IDX = 0

# Plotting options
CMAP = "RdBu_r"
DIFF_CMAP = "PiYG"
OUTPUT = "aifs_ens_compare.jpg"

old_tensor = torch.load(OLD_PT, map_location="cpu")
mk_tensor  = torch.load(MK_PT, map_location="cpu")

old_np = old_tensor.detach().cpu().numpy()
mk_np  = mk_tensor.detach().cpu().numpy()

# Load variable vocabularies
old_vars = np.load(OLD_VARS, allow_pickle=True)
mk_vars  = np.load(MK_VARS, allow_pickle=True)

# Ensure they are lists of strings
old_vars = [str(v) for v in old_vars.tolist()]
mk_vars  = [str(v) for v in mk_vars.tolist()]
common_vars = [v for v in mk_vars if v in old_vars]

# Map variable name to index
old_index = {v: i for i, v in enumerate(old_vars)}
mk_index  = {v: i for i, v in enumerate(mk_vars)}

T_old, L_old, _, Y_old, X_old = old_np.shape
T_mk,  L_mk,  _, Y_mk,  X_mk  = mk_np.shape


n = len(common_vars)
ncols = 3
nrows = n
fig, axs = plt.subplots(nrows=nrows, ncols=ncols, figsize=(12, 3.5 * nrows))

# Ensure axs is 2D
axs = np.atleast_2d(axs)

for i, v in enumerate(common_vars):
    ov = old_np[TIME_IDX, LEAD_IDX, old_index[v], :, :]
    mv = mk_np[TIME_IDX, LEAD_IDX, mk_index[v],  :, :]

    # Robust color scale for the two fields together
    stack = np.concatenate([ov.ravel(), mv.ravel()])
    stack = stack[np.isfinite(stack)]
    if stack.size > 0:
        vmin, vmax = np.nanpercentile(stack, [2, 98])
    else:
        vmin, vmax = np.nanmin(ov), np.nanmax(ov)

    # Diff symmetric scale
    diff = mv - ov
    dstack = diff[np.isfinite(diff)]
    if dstack.size > 0:
        dmin, dmax = np.nanpercentile(dstack, [2, 98])
        dabs = max(abs(dmin), abs(dmax))
    else:
        dabs = max(abs(np.nanmin(diff)), abs(np.nanmax(diff)))

    # OLD
    ax = axs[i, 0]
    im0 = ax.imshow(ov, cmap=CMAP, vmin=vmin, vmax=vmax, origin="upper")
    ax.set_title(f"OLD {v} [t={TIME_IDX}, l={LEAD_IDX}]")
    ax.set_xticks([]); ax.set_yticks([])
    fig.colorbar(im0, ax=ax, orientation="horizontal", fraction=0.05, pad=0.06)

    # MK
    ax = axs[i, 1]
    im1 = ax.imshow(mv, cmap=CMAP, vmin=vmin, vmax=vmax, origin="upper")
    ax.set_title(f"MK {v} [t={TIME_IDX}, l={LEAD_IDX}]")
    ax.set_xticks([]); ax.set_yticks([])
    fig.colorbar(im1, ax=ax, orientation="horizontal", fraction=0.05, pad=0.06)

    # DIFF
    ax = axs[i, 2]
    im2 = ax.imshow(diff, cmap="PiYG", vmin=-dabs, vmax=dabs, origin="upper")
    ax.set_title(f"Δ (MK − OLD) {v}")
    ax.set_xticks([]); ax.set_yticks([])
    fig.colorbar(im2, ax=ax, orientation="horizontal", fraction=0.05, pad=0.06)

plt.tight_layout()
plt.savefig("aifs_ens_compare_vars_test.jpg", dpi=150)
```

</details>

Results are not 100% identical, there is likely some fp16 error or some other randomness not fully set in the model but they are under 1% difference which seems accurate.

<details>

<summary> Comparison </summary>

aifs_ens_compare_vars_test
aifs_ens_compare_vars_test_2

</details>


Coverage is excellent for this model:

```
earth2studio/models/px/aifsens.py                   258      1    99%   327
```

File: earth2studio/models/px/aifsens.py [hunk 1]
```diff
@@ -228,33 +255,27 @@ def input_variables(self) -> list[str]:
         # Sort the concatenated tensor
         indices = indices.sort().values
 
-        # Create the range of values to remove
```
Comment (loliverhennigh) [inline earth2studio/models/px/aifsens.py:231]: This is awesome. Would be good to do this with AIFS as well. I think I can do this with the AIFS issue I am fixing.

**Review by loliverhennigh — APPROVED** [review]
LGTM

File: earth2studio/models/px/aifsens.py (hunk 1 above)
Comment (NickGeneva) [reply earth2studio/models/px/aifsens.py:231]: Makes this model a lot easier to run. Thanks for the review!
